### PR TITLE
trace debugging method

### PR DESCRIPTION
**Pull request by karlthepagan**
Implementation for #480

example traces: https://github.com/ScreepsOCS/screeps.behaviour-action-pattern/tree/debuggingWithTraces (will need rebase, so I'll push when this is ready)

### CODE REVIEW

File: global.js [hunk 1]
```diff
@@ -1,3 +1,18 @@
+const reduceMemoryWhere = function(result, value, key) {
+    const setting = Memory.debugTrace[key];
+    if (setting === undefined) {
+        return result;
+    } else if (result) { // default result
+        return setting === value;
```
Comment (karlthepagan) [inline global.js]: all parameters present must match and there must be at least one

**Comment by karlthepagan** [comment]
Example for CLI:

```
Memory.debugTrace.creep = true; // displays all actions in the "creep" category (action & behaviour selection)
Memory.debugTrace.no = {creepName:'hauler-1200-1'}; // ignores this creep
```

There's a lot of logic that I avoided implementing in the "where clause" logic. Consider this an exercise to the user to add callbacks or regex's.

File: global.js [hunk 2]
```diff
@@ -168,6 +183,13 @@ var mod = {
                 let text = dye(CRAYON.system, roomName);
                 console.log( dye(CRAYON.system, `<a href="/a/#!/room/${roomName}">${text}</a> &gt; `) + message );
             },
+            trace: function(category, entityWhere, ...message) {
+                if(! DEBUG) return;
```
Comment (cyberblast) [inline global.js]: hmm I guess you mean `if(!TRACE)` ? or maybe `!(DEBUG&&TRACE)` ?

Comment (karlthepagan) [inline global.js]: In the pattern that I'm using, I guard TRACE outside the method (preventing object creation), but I guess it wouldn't hurt to do it inside.

Comment (karlthepagan) [inline global.js]: @cyberblast after seeing some more error conditions I think that trace should work as-is. We can use the same function for exceptional case logging, and the practice writing these categories will hopefully translate into more trace logging written.

Comment (karlthepagan) [inline global.js]: However I see that overloading the word trace is going to cause friction so I'll add 2 logging methods, one for error conditions and one for traces.

**Comment by karlthepagan** [comment]
Adding trace log conventions will allow users to temporarily suppress spammy error conditions.

**Comment by karlthepagan** [comment]
rebased onto the refactor

**Pull request merged**